python/ray/data/tests/test_optimize.py

```diff
@@ -17,6 +19,70 @@ def expect_stages(pipe, num_stages_expected, stage_names):
     assert len(pipe._optimized_stages) == num_stages_expected, pipe._optimized_stages
 
 
+def test_memory_sanity(shutdown_only):
+    info = ray.init(num_cpus=1, object_store_memory=500e6)
+    ds = ray.data.range(10)
+    ds = ds.map(lambda x: np.ones(100 * 1024 * 1024, dtype=np.uint8))
+    meminfo = memory_summary(info.address_info["address"], stats_only=True)
+
+    # Sanity check spilling is happening as expected.
+    assert "Spilled" in meminfo, meminfo
+
+
+def test_memory_release_eager(shutdown_only):
+    info = ray.init(num_cpus=1, object_store_memory=1500e6)
+    ds = ray.data.range(10)
+
+    # Round 1.
+    ds = ds.map(lambda x: np.ones(100 * 1024 * 1024, dtype=np.uint8))
+    meminfo = memory_summary(info.address_info["address"], stats_only=True)
+    assert "Spilled" not in meminfo, meminfo
+
+    # Round 2.
+    ds = ds.map(lambda x: np.ones(100 * 1024 * 1024, dtype=np.uint8))
+    meminfo = memory_summary(info["address"], stats_only=True)
+    # TODO(ekl) we can add this back once we clear inputs on eager exec as well.
+    # assert "Spilled" not in meminfo, meminfo
+
+
+def test_memory_release_lazy(shutdown_only):
+    info = ray.init(num_cpus=1, object_store_memory=1500e6)
+    ds = ray.data.range(10)
+
+    # Should get fused into single stage.
+    ds = ds._experimental_lazy()
+    ds = ds.map(lambda x: np.ones(100 * 1024 * 1024, dtype=np.uint8))
+    ds = ds.map(lambda x: np.ones(100 * 1024 * 1024, dtype=np.uint8))
+    ds = ds.map(lambda x: np.ones(100 * 1024 * 1024, dtype=np.uint8))
+    ds.fully_executed()
+    meminfo = memory_summary(info.address_info["address"], stats_only=True)
+    assert "Spilled" not in meminfo, meminfo
+
+
+def test_memory_release_lazy_shuffle(shutdown_only):
+    # TODO(ekl) why is this flaky? Due to eviction delay?
```

Comment:
FYI @stephanie-wang , not sure if you also observed this in your shuffle experiments (memory sometimes not getting released in a timely manner during shuffle)

Reply:
Hmm I think @franklsf95 actually did see something like this.

Reply:
It's quite odd. I don't think it's related to Python GC or anything. I figured it might be a unit-test artifact since we have some batching/delay before releasing memory, which would only show up in a fast test, but it could also be an issue at larger scale.